accepted/future-releases/unified-collections/feature-specification.md

```diff
@@ -0,0 +1,802 @@
+        type, and one element which has only a key/value type, then it is a
+        compile-time error since the literal is constrained to be both a map and
+        a set.
+
+#### Lists
+
+Inference for list literals mostly follows that of set literals without the
+complexity around disambiguation with maps and using `List` or `Iterable` in a
+few places instead of `Set`.
+
+Inside a `listLiteral`, the inferred type of an `element` is a list element type
+`T`. It is computed relative to a context type `P`:
+
+*   If `element` is an `expressionElement` with expression `e1`:
+
+    *   If `P` is `?` then the inferred list element type of `element` is the
+        inferred type of the expression `e1` in context `?`.
+
+    *   If `P` is `Iterable<Ps>` then the inferred list element type of
+        `element` is the inferred type of the expression `e1` in context `Ps`.
+
+*   If `element` is a `spreadElement` with expression `e1`:
+
+    *   If `P` is `?` then let `S` be the inferred type of `e1` in context `?`:
+
+        *   If `S` is a subtype of `Iterable<Object>`, then the inferred list
+            element type of `element` is `T` where `T` is the type such that
+            `Iterable<T>` is a superinterface of `S` (the result of constraint
+            matching for `X` using the constraint `S <: Iterable<X>`).
+
+        *   If `S` is `dynamic`, then the inferred list element type of
+            `element` is `dynamic`.
+
+        *   Otherwise it is an error (because it is a spread of a non-spreadable
+            type like Object).
+
+    *   If `P` is `Iterable<Ps>` then let `S` be the inferred type of `e1` in
+        context `Iterable<Ps>`:
+
+        *   If `S` is a non-`Null` subtype of `Iterable<Object>`, then the
+            inferred list element type of `element` is `T` where `T` is the type
+            such that `Iterable<T>` is a superinterface of `S` (the result of
+            constraint matching for `X` using the constraint `Iterable<X> <:
+            S`).
+
+        *   If `S` is `dynamic`, then the inferred list element type of
+            `element` is `dynamic`.
+
+        *   Otherwise it is an error.
+
+*   If `element` is an `ifElement` with one `element`, `p1`, and no `else`:
+
+    The condition is inferred with a context type of `bool`.
+
+    The inferred list element type of `element` is the inferred list element
+    type of `p1`.
+
+*   If `element` is a `ifElement` with two `element`s, `p1` and `p2`:
+
+    The condition is inferred with a context type of `bool`.
+
+    The inferred list element type of `element` is the upper bound of `S1` and
+    `S2` where `S1` is the inferred list element type of `p1` and `S2` is the
+    inferred list element type of `p2`.
+
+*   If `element` is a `forElement` with `element` `p1` then:
+
+    Inference for the iterated expression and the controlling variable is done
+    as for the corresponding `for` or `await for` statement.
```

Comment:
Are we sure we don't need *more* that we do for statements when we have a context type for the body?
Example: 
```dart
<int>{for (var x in listOfObject.whereType()) x}
```
Here there is an expected type of `int` for `x`, so we could technically let that flow back into the loop-parts and infer `.whereType<int>()`.

We would do that inference for `<int>{...objectList.whereType()}`.

It might be too complicated, and not really different from statements because they can introduce constraints on the loop variable as well, by using it.
Just want to bring up the issue in case there is something we should do more.

Reply:
It feels a little magical to flow backwards from the element body to the loop variable and then to the iterator expression from there. It makes sense in the example here because the body is literally just the loop variable, but that's a degenerate case. No one should write that when they can instead do:

```dart
<int>{...listOfObject.whereType())}
```

In realistic cases where the body of a for element is a more complex expression, I don't know if we could easily flow inference through it. I'll defer to Leaf, but I think what we have is sufficient.